packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -695,136 +700,317 @@ class _SortData {
+  /// based on an explicit order.
+  ///
+  /// If [secondary] is null, it will default to [ReadingOrderTraversalPolicy].
+  OrderedFocusTraversalPolicy({this.secondary}) {
+    secondary ??= ReadingOrderTraversalPolicy();
+  }
+
+  /// This is the policy that is used when a node doesn't have an order
+  /// assigned, or when multiple nodes have orders which are identical.
+  ///
+  /// If not set, this defaults to [ReadingOrderTraversalPolicy].
+  ///
+  /// This policy determines the secondary sorting order of nodes which evaluate
+  /// as having an identical order (including those with no order specified).
+  FocusTraversalPolicy secondary;
+
+  @override
+  Iterable<FocusNode> sortDescendants(FocusScopeNode scope) {
+    final Iterable<FocusNode> geometryOrdered = secondary.sortDescendants(scope);
+    final List<FocusNode> unordered = <FocusNode>[];
+    final List<_OrderedFocusInfo> ordered = <_OrderedFocusInfo>[];
+    for (final FocusNode node in geometryOrdered) {
+      final FocusOrder order = FocusTraversalOrder.of(node.context, nullOk: true);
+      if (order != null) {
+        ordered.add(_OrderedFocusInfo(node: node, order: order));
+      } else {
+        unordered.add(node);
       }
     }
-    final List<FocusNode> sortedNodes = _sortByGeometry(nearestScope).toList();
-    if (forward && focusedChild == sortedNodes.last) {
-      _focusAndEnsureVisible(sortedNodes.first, alignmentPolicy: ScrollPositionAlignmentPolicy.keepVisibleAtEnd);
-      return true;
-    }
-    if (!forward && focusedChild == sortedNodes.first) {
-      _focusAndEnsureVisible(sortedNodes.last, alignmentPolicy: ScrollPositionAlignmentPolicy.keepVisibleAtStart);
-      return true;
-    }
+    ordered.sort((_OrderedFocusInfo a, _OrderedFocusInfo b) {
+      return a.order.compareTo(b.order);
+    });
+    return ordered.map<FocusNode>((_OrderedFocusInfo info) => info.node).followedBy(unordered);
+  }
+}
 
-    final Iterable<FocusNode> maybeFlipped = forward ? sortedNodes : sortedNodes.reversed;
-    FocusNode previousNode;
-    for (final FocusNode node in maybeFlipped) {
-      if (previousNode == focusedChild) {
-        _focusAndEnsureVisible(
-          node,
-          alignmentPolicy: forward
-            ? ScrollPositionAlignmentPolicy.keepVisibleAtEnd
-            : ScrollPositionAlignmentPolicy.keepVisibleAtStart,
+/// An inherited widget that describes the order in which its child subtree
+/// should be traversed.
+///
+/// The order for a widget is determined by the [FocusOrder] returned by
+/// [FocusTraversalOrder.of] for a particular context.
+class FocusTraversalOrder extends InheritedWidget {
+  /// A const constructor so that subclasses can be const.
+  const FocusTraversalOrder({Key key, this.order, Widget child}) : super(key: key, child: child);
+
+  /// The order for the widget descendants of this [FocusTraversalOrder].
+  final FocusOrder order;
+
+  /// Finds the nearest TraversalOrder widget.
```

Comment:
It seems to return a FocusOrder though instead of a TraversalOrder widget. Also, a TraversalOrder widget doesn't exist.